system/lib/libc/dynlink.c

```diff
@@ -586,35 +586,11 @@ void* dlopen(const char* file, int flags) {
 
 void emscripten_dlopen(const char* filename, int flags, void* user_data,
                        em_dlopen_callback onsuccess, em_arg_callback_func onerror) {
-  dbg("emscripten_dlopen: %s", filename);
-  if (!filename) {
-    onsuccess(user_data, head->dso);
-    return;
-  }
-  do_write_lock();
-  char buf[2*NAME_MAX+2];
-  filename = resolve_path(buf, filename, sizeof buf);
-  struct dso* p = find_existing(filename);
-  if (p) {
-    onsuccess(user_data, p);
-    return;
-  }
-  p = load_library_start(filename, flags);
-  if (!p) {
-    do_write_unlock();
-    onerror(user_data);
-    return;
-  }
-
-  // For async mode
-  struct async_data* d = malloc(sizeof(struct async_data));
-  d->user_data = user_data;
-  d->onsuccess = onsuccess;
-  d->onerror = onerror;
+  
+  em_promise_t promise = emscripten_dlopen_promise(filename,flags);
+  emscripten_promise_then(promise, onsuccess, user_data);
+  emscripten_promise_catch(promise, onerror, user_data);
```

Comment:
This nice, it seems like you have implemented something like `promiseToCallback` in native code?

I think it might be fast it if was implemented JS though since using promises from the native does have some amount of overhead I think (just calling back and forth into JS has a cost).